InternalEngineTests testRecoverFromHardDeletesIndex failure

```
org.elasticsearch.index.engine.InternalEngineTests > testRecoverFromHardDeletesIndex FAILED
    java.lang.AssertionError: Operation not found seq# [16], global checkpoint [59], retention policy [0], maxSeqNo [65], translog op [Index{id='12', seqNo=16, primaryTerm=647288508455157647, version=16, autoGeneratedIdTimestamp=-1}]
        at __randomizedtesting.SeedInfo.seed([C54FB26E3EA917B1:101D892A1EE75753]:0)
        at org.junit.Assert.fail(Assert.java:88)
        at org.elasticsearch.index.engine.EngineTestCase.assertConsistentHistoryBetweenTranslogAndLuceneIndex(EngineTestCase.java:1105)
        at org.elasticsearch.index.engine.InternalEngineTests.testRecoverFromHardDeletesIndex(InternalEngineTests.java:5489)
```

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=oraclelinux-7&&immutable/496/console
https://gradle-enterprise.elastic.co/s/fk25he4qu3rye

Another win for reproducibility:
```
./gradlew ':server:test' --tests "org.elasticsearch.index.engine.InternalEngineTests.testRecoverFromHardDeletesIndex" \
  -Dtests.seed=C54FB26E3EA917B1 \
  -Dtests.security.manager=true \
  -Dtests.locale=fr \
  -Dtests.timezone=Indian/Antananarivo \
  -Dcompiler.java=13
```

Possibly related to #50517

--- a/server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java
+++ b/server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java
@@ -5476,10 +5476,14 @@ public class InternalEngineTests extends EngineTestCase {
         final AtomicLong globalCheckpoint = new AtomicLong(SequenceNumbers.NO_OPS_PERFORMED);
         Path translogPath = createTempDir();
         List<Engine.Operation> operations = generateHistoryOnReplica(between(1, 500), randomBoolean(), randomBoolean(), randomBoolean());
+        final IndexMetaData indexMetaData = IndexMetaData.builder(defaultSettings.getIndexMetaData())
+            .settings(Settings.builder().put(defaultSettings.getSettings()).put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), false))
+            .build();
+        final IndexSettings indexSettings = IndexSettingsModule.newIndexSettings(indexMetaData);
         try (Store store = createStore()) {
-            EngineConfig config = config(defaultSettings, store, translogPath, NoMergePolicy.INSTANCE, null, null, globalCheckpoint::get);
+            EngineConfig config = config(indexSettings, store, translogPath, NoMergePolicy.INSTANCE, null, null, globalCheckpoint::get);
             final List<DocIdSeqNoAndSource> docs;
-            try (InternalEngine hardDeletesEngine = createEngine(defaultSettings, store, translogPath, newMergePolicy(),
+            try (InternalEngine hardDeletesEngine = createEngine(indexSettings, store, translogPath, newMergePolicy(),
                 hardDeletesWriter, null, globalCheckpoint::get)) {
                 for (Engine.Operation op : operations) {
                     applyOperation(hardDeletesEngine, op);

